incubator/validator-avro/src/main/java/io/aklivity/zilla/runtime/validator/avro/AvroReadValueValidator.java

```diff
@@ -0,0 +1,132 @@
+        ToLongFunction<String> resolveId,
+        LongFunction<CatalogHandler> supplyCatalog)
+    {
+        super(config, resolveId, supplyCatalog);
+    }
+
+    @Override
+    public int validate(
+        DirectBuffer data,
+        int index,
+        int length,
+        ValueConsumer next)
+    {
+        MutableDirectBuffer value = new UnsafeBuffer();
+        int valLength = -1;
+
+        byte[] payloadBytes = new byte[length];
+        data.getBytes(0, payloadBytes);
+        ByteBuffer byteBuf = ByteBuffer.wrap(payloadBytes);
+
+        int schemaId;
+        Schema schema;
+        if (byteBuf.get() == MAGIC_BYTE)
+        {
+            schemaId = byteBuf.getInt();
+            int size = length - 1 - 4;
+            byte[] valBytes = new byte[size];
+            data.getBytes(length - size, valBytes);
+            schema = fetchSchema(schemaId);
+            if (schema != null)
+            {
+                if ("json".equals(expect))
+                {
+                    byte[] record = serializeAvroRecord(schema, valBytes);
+                    value.wrap(record);
+                    valLength = record.length;
+                }
+                else if (validate(schema, valBytes))
+                {
+                    value.wrap(data);
+                    valLength = length;
+                }
+            }
+        }
+        else
+        {
+            schemaId = catalog != null ? catalog.id : 0;
+            schema = fetchSchema(schemaId);
+            if (schema != null)
+            {
+                if ("json".equals(expect))
+                {
+                    byte[] record = serializeAvroRecord(schema, payloadBytes);
+                    value.wrap(record);
+                    valLength = record.length;
+                }
+                else if (validate(schema, payloadBytes))
+                {
+                    value.wrap(data);
+                    valLength = length;
+                }
+            }
+        }
+
+        next.accept(value, index, valLength);
+        return valLength;
+    }
+
+    private byte[] serializeAvroRecord(
+        Schema schema,
+        byte[] payloadBytes)
+    {
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        try
+        {
+            reader = new GenericDatumReader(schema);
+            GenericRecord record = (GenericRecord) reader.read(null,
+                    decoder.binaryDecoder(payloadBytes, null));
+            JsonEncoder jsonEncoder = encoder.jsonEncoder(record.getSchema(), outputStream);
+            writer = record instanceof SpecificRecord ?
+                new SpecificDatumWriter<>(record.getSchema()) :
+                new GenericDatumWriter<>(record.getSchema());
+            writer.write(record, jsonEncoder);
+            jsonEncoder.flush();
+            outputStream.close();
+        }
+        catch (Exception e)
+        {
+        }
+        return outputStream.toByteArray();
+    }
```

Comment:
Lets aim to avoid repeated work per invocation that can be hoisted and done just once instead, then reused.